backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/VdsCpuUnitPinningHelper.java

```diff
@@ -129,64 +112,60 @@ private void previewPinOfPendingExclusiveCpus(List<VdsCpuUnit> cpuTopology, Map<
-                              CpuPinningPolicy cpuPinningPolicy) {
-        for (int core : getOnlineCores(cpuTopology, socket)) {
-            List<VdsCpuUnit> sharedCpus = getFreeCpusInCore(cpuTopology, socket, core);
-            int sharedCpusInCore = sharedCpus.size();
-            List<Integer> sharedCpuIds = sharedCpus.stream().map(VdsCpuUnit::getCpu).collect(Collectors.toCollection(LinkedList::new));
-            Iterator<Integer> sharedCpusIterator = sharedCpuIds.iterator();
-            if (cpusLeft == 0) {
-                return cpusLeft;
-            }
-            if (wholeCore) {
-                if (cpusLeft >= hostThreadsPerCore && sharedCpusInCore == hostThreadsPerCore) {
-                    // take the whole core
-                    while (sharedCpusInCore-- > 0) {
-                        VdsCpuUnit cpuToTake = getCpu(sharedCpus, socket, core, sharedCpusIterator.next());
-                        cpusToBeAllocated.add(cpuToTake);
-                        cpuToTake.pinVm(vmId, cpuPinningPolicy);
-                        cpusLeft--;
+    private int allocateCpus(List<VdsCpuUnit> cpuTopology, VM vm, List<VdsCpuUnit> cpusToBeAllocated, boolean wholeCore) {
+        int socketsLeft = vm.getNumOfSockets();
+        int onlineSockets = getOnlineSockets(cpuTopology).size();
+        while (onlineSockets > 0 && socketsLeft > 0) {
+            List<VdsCpuUnit> currentSocket = getMaxFreedSocket(cpuTopology);
+            // coreCount is based on the VM topology
+            int coreCount = 0;
+            for (int core : getOnlineCores(cpuTopology)) {
+                List<VdsCpuUnit> freeCpusInCore = getFreeCpusInCore(currentSocket, core);
+                int coreThreads = freeCpusInCore.size();
+                int index = 0;
+                while (coreThreads >= vm.getThreadsPerCpu() && wholeCoreTaken(freeCpusInCore, vm.getThreadsPerCpu(), wholeCore)) {
+                    for (int thread = 0; thread < vm.getThreadsPerCpu(); thread++) {
+                        VdsCpuUnit cpuUnit = freeCpusInCore.get(thread + index);
+                        cpuUnit.pinVm(vm.getId(), vm.getCpuPinningPolicy());
+                        cpusToBeAllocated.add(cpuUnit);
                     }
+                    coreCount++;
+                    coreThreads -= vm.getThreadsPerCpu();
+                    index += vm.getThreadsPerCpu();
                 }
-            } else {
-                while (sharedCpusInCore-- > 0 && cpusLeft > 0) {
-                    VdsCpuUnit cpuToTake = getCpu(sharedCpus, socket, core, sharedCpusIterator.next());
-                    cpusToBeAllocated.add(cpuToTake);
-                    cpuToTake.pinVm(vmId, cpuPinningPolicy);
-                    cpusLeft--;
+            }
+            while (socketsLeft > 0 && coreCount >= vm.getCpuPerSocket()) {
+                socketsLeft--;
+                coreCount -= vm.getCpuPerSocket();
+            }
```

Comment:
it would be easier if you add a comment before such a block that describes its purpose